src/TestUtils.jl

```diff
@@ -0,0 +1,173 @@
+
+Run tests to verify correctness of FFT functions using a particular 
+backend plan implementation. The backend implementation is assumed to be loaded 
+prior to calling this function.
+
+# Arguments
+
+- `array_constructor`: determines the `AbstractArray` implementation for
+  which the correctness tests are run. It is assumed to be a callable object that
+  takes in input arrays of type `Array` and return arrays of the desired type for
+  testing. For example, this can be a constructor such as `Array` or `CUDA.CuArray`.
+- `test_real=true`: whether to test real-to-complex and complex-to-real FFTs.
+- `test_inplace=true`: whether to test in-place plans. 
+"""
+function test_fft_backend(array_constructor; test_real=true, test_inplace=true)
+    @testset "fft correctness" begin
+        # DFT along last dimension, results computed using FFTW
+        for (_x, dims, real_input, _fftw_fft) in (
+            (collect(1:7), 1, true,
+             [28.0 + 0.0im,
+              -3.5 + 7.267824888003178im,
+              -3.5 + 2.7911568610884143im,
+              -3.5 + 0.7988521603655248im,
+              -3.5 - 0.7988521603655248im,
+              -3.5 - 2.7911568610884143im,
+              -3.5 - 7.267824888003178im]),
+            (collect(1:8), 1, true,
+             [36.0 + 0.0im,
+              -4.0 + 9.65685424949238im,
+              -4.0 + 4.0im,
+              -4.0 + 1.6568542494923806im,
+              -4.0 + 0.0im,
+              -4.0 - 1.6568542494923806im,
+              -4.0 - 4.0im,
+              -4.0 - 9.65685424949238im]),
+            (collect(reshape(1:8, 2, 4)), 2, true,
+             [16.0+0.0im  -4.0+4.0im  -4.0+0.0im  -4.0-4.0im;
+              20.0+0.0im  -4.0+4.0im  -4.0+0.0im  -4.0-4.0im]),
+            (collect(reshape(1:9, 3, 3)), 2, true,
+             [12.0+0.0im  -4.5+2.598076211353316im  -4.5-2.598076211353316im;
+              15.0+0.0im  -4.5+2.598076211353316im  -4.5-2.598076211353316im;
+              18.0+0.0im  -4.5+2.598076211353316im  -4.5-2.598076211353316im]),
+            (collect(reshape(1:8, 2, 2, 2)), 1:2, true,
+             cat([10.0 + 0.0im -4.0 + 0.0im; -2.0 + 0.0im 0.0 + 0.0im],
+                 [26.0 + 0.0im -4.0 + 0.0im; -2.0 + 0.0im 0.0 + 0.0im],
+                 dims=3)),
+            (collect(1:7) + im * collect(8:14), 1, false,
+             [28.0 + 77.0im,
+              -10.76782488800318 + 3.767824888003175im,
+              -6.291156861088416 - 0.7088431389115883im,
+              -4.298852160365525 - 2.7011478396344746im,
+              -2.7011478396344764 - 4.298852160365524im,
+              -0.7088431389115866 - 6.291156861088417im,
+              3.767824888003177 - 10.76782488800318im]),
+            (collect(reshape(1:8, 2, 2, 2)) + im * reshape(9:16, 2, 2, 2), 1:2, false,
+             cat([10.0 + 42.0im -4.0 - 4.0im; -2.0 - 2.0im 0.0 + 0.0im],
+                 [26.0 + 58.0im -4.0 - 4.0im; -2.0 - 2.0im 0.0 + 0.0im],
+                 dims=3)),
```

Comment:
I wonder if there is a better, i.e., more readable way, to specify these test inputs and results. Also, is the third component (`false`/`true`) actually needed? Seems like that could be inferred from the type of `_x`?

Reply:
The best I could think of was to specify the test cases using named tuples so it is easier to keep track of what means what. And you're right about the third component, I've removed it.